### ISSUE TITLE
Bandwidth rate restriction cannot be applied after playback started.

### ISSUE THREAD

**Issue by SemihGk** [google/shaka-player#1533]
**Have you read the [FAQ](https://bit.ly/ShakaFAQ) and checked for duplicate open issues?**:
Yes
**What version of Shaka Player are you using?**:
latest master or 2.4.3
**Can you reproduce the issue with our latest release version?**:
yes
**Can you reproduce the issue with the latest code from `master`?**:
yes
**Are you using the demo app or your own custom app?**:
demo app
**If custom app, can you reproduce the issue using our demo app?**:

**What browser and OS are you using?**:
chrome and macOS
**What are the manifest and license server URIs?**:
Does not matter. You may check tears of steel(widevine) sample.

**What did you do?**
I did set max bandwidth rate after playback started. However, the restrictions cannot be applied to abrManager even though restrictions applied to configuration object.

For instance, call  before playback.
shakaDemo.localPlayer_.configure({ restrictions: { maxBandwidth: 500000 } });

Then play the content and call:
shakaDemo.localPlayer_.configure({ restrictions: { maxBandwidth: Infinity } });

**What did you expect to happen?**
The restriction should be applied to abrManager.

**What actually happened?**
The restriction could not be applied and abrManager only played variants by restricted of the previous value.

Note: The issue is fixed if I add this change at [this block](https://github.com/google/shaka-player/blob/master/lib/player.js#L1220)
```
    if (!activeVariant || !activeVariant.allowedByApplication ||
        !activeVariant.allowedByKeySystem) {
      shaka.log.debug('Choosing new streams after changing configuration');
      this.chooseStreamsAndSwitch_(period);
    } else {
      // If we change the configuration during the playback, the changes cannot
      // be applied due to abrManager is not updated. So, we force to update it.
      let variants = shaka.util.StreamUtils.filterVariantsByConfig(
          period.variants, this.currentAudioLanguage_, this.currentVariantRole_,
          this.currentAudioChannelCount_);
      this.abrManager_.setVariants(variants);
    }
```

I don't mind to create a PR if this change is appropriate for this fix. Thank you.

**Comment by chrisfillmore** [comment]
I'm also curious about the purpose of soft restrictions in https://github.com/google/shaka-player/commit/ed0f37b0990ea9a4a40e6c1779822794367ab7f8 . Why allow ABR to pick a variant if all are restricted?

**Comment by ismena** [comment]
@chrisfillmore We should, probably, rename them to preferences or something.
@joeyparrish, please correct me, if I'm wrong (I was not actively involved with this change), but, to the best my knowledge, the idea is that:
- top-level restrictions mean "we can/shall not play the variants that don't comply with the restrictrictions, treat them like they don't exist"
- abr restrictions (in the new world, the ones that I think should be called preferences) mean "we can/shall play these, but we'd rather not. If there is any other variant we can play instead, let's do it; if not, so be it".

The data savers thing references in the commit msg is basically "if the data saving mode is on, make every attempt to play lower bandwidth video before going for higher res and wasting precious data".

**Comment by joeyparrish** [comment]
It used to be the case (I think) that restrictions would be reapplied when the config changed.  So this may be a regression.

Also, to clarify an earlier discussion, "soft" ABR restrictions only apply to ABR logic.  Tracks restricted at that level can still be selected manually via `selectVariantTrack()`, so it is not redundant with "hard" restrictions.  The top-level "hard" restrictions are meant to communicate something that should not be played, and have a primary purpose of communicating out-of-band DRM restrictions.  For example, on IE11, where key statuses are not available, restrictions and out-of-band DRM hints are the only way to avoid playing tracks that can't be decrypted with a given license.

**Comment by joeyparrish** [comment]
I attempted to cherry-pick this bug fix to v2.4.5, but the merge conflicts were too complex to fix up.  So this bug fix will not appear in a release until v2.5.0-beta2.  We apologize for the inconvenience.

**Comment by joeyparrish** [comment]
Released in v2.5.0-beta2.
